src/components/datasource.vue

```diff
@@ -0,0 +1,452 @@
+            class="uk-button uk-button-primary"
+            @click="onSave">Save</button>
+        </div>
+      </div>
+    </form>
+  </div>
+</template>
+
+<script>
+import DataConnector from './dataconnector.vue'
+import { validationMixin } from 'vuelidate'
+import { required } from 'vuelidate/lib/validators'
+import nanoid from 'nanoid'
+import store from './store'
+import templayed from 'templayed'
+import { ArcGISLayer } from '../lib/arcgis'
+import { debounce } from 'lodash'
+
+export default {
+  name: 'data-source',
+  mixins: [validationMixin],
+  components: {
+    DataConnector,
+  },
+  props: {
+    uid: {
+      type: String,
+      default () { return nanoid() },
+    },
+    dataSourceType: {
+      type: String,
+      default () { return store.state.dataSourceTypeChoices['CoB ArcGIS'] }
+    },
+    dataSourceTypeChoices: {
+      type: Object,
+      // default () { return {'CoB ArcGIS': 'cob-arcgis'} }
+      default () {
+        return store.state.dataSourceTypeChoices
+      }
+    },
+    attributes: {
+      type: Object,
+      default () { return {} }
+    },
+    icon: {
+      type: String,
+      // default () { return this.iconChoices.default }
+      default () { return store.state.iconChoices.default }
+    },
+    iconChoices: {
+      type: Object,
+      default () {
+        return store.state.iconChoices
+      }
+    },
+    clusterPoints: {
+      type: Boolean,
+      default () { return true },
+    },
+    polygonStyle: {
+      type: String,
+      default () { return 'default' }
+    },
+    polygonStyleChoices: {
+      type: Object,
+      default () { return store.state.polygonStyleChoices }
+    },
+    popover: {
+      type: String,
+      default () { return '' }
+    },
+    legendLabel: {
+      type: String,
+      default () { return '' }
+    },
+  },
+  data () {
+    return {
+      selectedDataSourceType: '',
+      selectedAttributes: {},
+      selectedIcon: '',
+      selectedClusterPoints: true,
+      selectedPolygonStyle: '',
+      selectedPopover: '',
+      selectedLegendLabel: '',
+      fields: [],
+      validateNow: false,
+      sampleRow: 1,
+      datasourceConnector: null,
+      currentRowCache: {},
+      // samplePopover: '',
+    }
+  },
+  computed: {
+    samplePopover: function () {
+      return templayed(this.selectedPopover)(this.currentRowCache)
+    },
+    configObject () {
+      let c = {
+        uid: this.uid,
+        type: this.selectedDataSourceType,
+        icon: this.selectedIcon,
+        clusterIcons: this.selectedClusterPoints,
+        polygonStyle: this.selectedPolygonStyle,
+        popover: this.selectedPopover,
+        attributes: Object.assign({}, this.selectedAttributes),
+        legendLabel: this.selectedLegendLabel,
+      }
+      return c
+    },
+    editMode () { return 'Edit' },
+    label () {
+      if (this.selectedLegendLabel) {
+        return this.selectedLegendLabel
+      }
+      let attrs = []
```

Comment:
This doesn't make sense to me. Why would I label a variable `const` if I am going to be adding to it? It isn't constant.

Reply:
`const` refers to whether you’re re-assigning the name `attrs` to a different value, not whether the value it refers to is mutable or not. Since `attrs` doesn’t get changed from this array, it should be `const`.